repentogon/Patches/ASMPatches/ASMStatusEffects.cpp

```diff
@@ -0,0 +1,169 @@
+
+	{"8bf7e8????????84c074??8b8f????????85c974??8bf18b8e????????85c975??8b8e",
+	"AddShrink", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x21},
+	{"e8????????84c074??8b8e????????85c974??0f1f00",
+	"AddSlowing", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x22},
+	{"e8????????84c074??8b8e????????85c974??8bf1",
+	"AddWeakness", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x1f},
+};
+
+/* /////////////////////
+// shared GetStatusEffectTarget trampoline
+*/ /////////////////////
+
+Entity* __stdcall GetStatusEffectTargetTrampoline(Entity* target) {
+	return target->GetStatusEffectTarget();
+}
+
+/* /////////////////////
+// Generic inline patch
+*/ /////////////////////
+
+void ASMPatchInlinedGetStatusEffectTarget(void* addr, ASMPatch::Registers entityReg, ASMPatch::Registers targetReg, ASMPatch::SavedRegisters::Registers savedReg, unsigned int jumpOffset) {
+	ASMPatch::SavedRegisters savedRegisters(ASMPatch::SavedRegisters::Registers::GP_REGISTERS + ASMPatch::SavedRegisters::XMM1 - savedReg, true);
+	ASMPatch patch;
+	patch.PreserveRegisters(savedRegisters)
+		.Push(entityReg)
+		.AddInternalCall(GetStatusEffectTargetTrampoline)
+		// poor man's mov reg,reg
+		.Push(ASMPatch::Registers::EAX)
+		.Pop(targetReg)
+		.RestoreRegisters(savedRegisters)
+		.AddRelativeJump((char*)addr + jumpOffset);
+	sASMPatcher.PatchAt(addr, &patch);
+}
+
+/* /////////////////////
+// Thank you, God bless you, and God bless the United States of America.
+*/ /////////////////////
+
+bool avoidIceCrash = false;
+void __stdcall AvoidIceCrashHack() {
+	avoidIceCrash = true;
+}
+
+void ASMPatchIcedKnockbackPreNPC() {
+	SigScan scanner("6a0168c80000008d85");
+	scanner.Scan();
+	void* addr = scanner.GetAddress();
+
+	ASMPatch patch;
+	patch.AddInternalCall(AvoidIceCrashHack)
+		.AddBytes("\x6a\x01\x68\xc8").AddZeroes(3)
+		.AddRelativeJump((char*)addr + 7);
+	sASMPatcher.PatchAt(addr, &patch);
+}
+
+void ASMPatchIcedKnockbackPrePlayer() {
+	SigScan scanner("68c8000000508d4424");
+	scanner.Scan();
+	void* addr = scanner.GetAddress();
+
+	ASMPatch patch;
+	patch.AddInternalCall(AvoidIceCrashHack)
+		.AddBytes("\x68\xc8").AddZeroes(3)
+		.AddRelativeJump((char*)addr + 5);
+	sASMPatcher.PatchAt(addr, &patch);
+}
+
+HOOK_METHOD(Entity, AddKnockback, (EntityRef& ref, const Vector& pushDirection, int duration, bool takeImpactDamage) -> void) {
+	if (!avoidIceCrash)
+		super(ref, pushDirection, duration, takeImpactDamage);
+	else
+	{
+		Entity* entity = this;
+		if (!IgnoreEffectFromFriendly(&ref)) {
+			if (GetStatusEffectTarget()) {
+				Entity* parent = this->_parent;
+				Entity* otherEnt = parent;
+				while (otherEnt = parent, otherEnt != nullptr) {
+					entity = otherEnt;
+					parent = otherEnt->_parent;
+				}
+			}
+			if (duration != 0 && ((entity->_flags & 1) == 0 || (entity->_flags & 0x2000000000000) != 0)) {
```

Comment:
Please add 0x2000000000000 as a define somewhere, I have no idea what it represents. Sames goes for the 1 that you use in the first part of the if

Reply:
What we really need is to port over our Ghidra enums. It doesn't seem like there's an immediate way to export them, though...

Reply:
Resolved by virtue of no longer reimplementing AddKnockback.